File: src/core/api/file.ts
/**
 * File API - Common file operations
 */
import { invoke } from '@tauri-apps/api/core';
import type { ApiResponse } from './types';

/**
 * File or directory info
 */
export interface FileInfo {
  name: string;
  path: string;
  is_dir: boolean;
  size: number;
  modified: string;
}

/**
 * Read text file content
 */
export async function readTextFile(filePath: string): Promise<ApiResponse<string>> {
  try {
    return await invoke<ApiResponse<string>>('read_text_file', { path: filePath });
  } catch (error) {
    console.error('[API] readTextFile error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Read binary file as base64
 */
export async function readBinaryFile(filePath: string): Promise<ApiResponse<string>> {
  try {
    return await invoke<ApiResponse<string>>('read_binary_file', { path: filePath });
  } catch (error) {
    console.error('[API] readBinaryFile error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Write text to file
 */
export async function writeTextFile(filePath: string, content: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('write_text_file', { path: filePath, content });
  } catch (error) {
    console.error('[API] writeTextFile error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Delete a file
 */
export async function deleteFile(filePath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('delete_file', { path: filePath });
  } catch (error) {
    console.error('[API] deleteFile error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Rename a file or directory
 */
export async function renamePath(oldPath: string, newPath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('rename_path', { old_path: oldPath, new_path: newPath });
  } catch (error) {
    console.error('[API] renamePath error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Create a directory
 */
export async function createDirectory(dirPath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('create_directory', { path: dirPath });
  } catch (error) {
    console.error('[API] createDirectory error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Delete a directory
 */
export async function deleteDirectory(dirPath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('delete_directory', { path: dirPath });
  } catch (error) {
    console.error('[API] deleteDirectory error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * List directory contents
 */
export async function listDirectory(dirPath: string): Promise<ApiResponse<FileInfo[]>> {
  try {
    return await invoke<ApiResponse<FileInfo[]>>('list_directory', { path: dirPath });
  } catch (error) {
    console.error('[API] listDirectory error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Copy a file
 */
export async function copyFile(sourcePath: string, destPath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('copy_file', { source: sourcePath, dest: destPath });
  } catch (error) {
    console.error('[API] copyFile error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Check if path exists
 */
export async function pathExists(filePath: string): Promise<ApiResponse<boolean>> {
  try {
    return await invoke<ApiResponse<boolean>>('path_exists', { path: filePath });
  } catch (error) {
    console.error('[API] pathExists error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * File change event from watcher
 */
export interface FileChangeEvent {
  path: string;
  kind: 'create' | 'modify' | 'remove';
  parent: string;
}

/**
 * Start watching a directory for file changes
 */
export async function startWatch(dirPath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('start_watch', { path: dirPath });
  } catch (error) {
    console.error('[API] startWatch error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Stop watching a directory
 */
export async function stopWatch(dirPath: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('stop_watch', { path: dirPath });
  } catch (error) {
    console.error('[API] stopWatch error:', error);
    return { success: false, error: String(error) };
  }
}

File: src-tauri/src/core/commands/file.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::time::SystemTime;

/// File or directory info
#[derive(Debug, Serialize, Deserialize)]
pub struct FileInfo {
    pub name: String,
    pub path: String,
    pub is_dir: bool,
    pub size: u64,
    pub modified: String,
}

/// Generic API response
#[derive(Debug, Serialize)]
pub struct ApiResponse<T> {
    pub success: bool,
    pub data: Option<T>,
    pub error: Option<String>,
}

impl<T> ApiResponse<T> {
    pub fn ok(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
        }
    }

    pub fn err(msg: String) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(msg),
        }
    }
}

fn system_time_to_string(time: SystemTime) -> String {
    let duration = time
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap_or_default();
    let secs = duration.as_secs();
    format!("{}", secs)
}

/// Read text file content
#[tauri::command]
pub async fn read_text_file(path: String) -> Result<ApiResponse<String>, String> {
    println!("[read_text_file] Received path: {:?}", path);
    match fs::read_to_string(&path) {
        Ok(content) => Ok(ApiResponse::ok(content)),
        Err(e) => Ok(ApiResponse::err(format!("读取文件失败: {} (path: {:?})", e, path))),
    }
}

/// Read binary file as base64
#[tauri::command]
pub async fn read_binary_file(path: String) -> Result<ApiResponse<String>, String> {
    use base64::{engine::general_purpose::STANDARD, Engine};
    println!("[read_binary_file] Received path: {:?}", path);

    match fs::read(&path) {
        Ok(bytes) => {
            let encoded = STANDARD.encode(&bytes);
            Ok(ApiResponse::ok(encoded))
        }
        Err(e) => Ok(ApiResponse::err(format!("读取文件失败: {} (path: {:?})", e, path))),
    }
}

/// Write text to file
#[tauri::command]
pub async fn write_text_file(path: String, content: String) -> Result<ApiResponse<()>, String> {
    match fs::write(&path, &content) {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(format!("写入文件失败: {}", e))),
    }
}

/// Delete a file
#[tauri::command]
pub async fn delete_file(path: String) -> Result<ApiResponse<()>, String> {
    match fs::remove_file(&path) {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(format!("删除文件失败: {}", e))),
    }
}

/// Rename a file or directory
#[tauri::command]
pub async fn rename_path(old_path: String, new_path: String) -> Result<ApiResponse<()>, String> {
    match fs::rename(&old_path, &new_path) {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(format!("重命名失败: {}", e))),
    }
}

/// Create a directory
#[tauri::command]
pub async fn create_directory(path: String) -> Result<ApiResponse<()>, String> {
    match fs::create_dir_all(&path) {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(format!("创建目录失败: {}", e))),
    }
}

/// Delete a directory
#[tauri::command]
pub async fn delete_directory(path: String) -> Result<ApiResponse<()>, String> {
    match fs::remove_dir_all(&path) {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(format!("删除目录失败: {}", e))),
    }
}

/// List directory contents
#[tauri::command]
pub async fn list_directory(path: String) -> Result<ApiResponse<Vec<FileInfo>>, String> {
    println!("[list_directory] Received path: {:?}", path);
    let entries = match fs::read_dir(&path) {
        Ok(entries) => entries,
        Err(e) => return Ok(ApiResponse::err(format!("读取目录失败: {}", e))),
    };

    let mut files = Vec::new();
    for entry in entries {
        if let Ok(entry) = entry {
            let path = entry.path();
            let metadata = match entry.metadata() {
                Ok(m) => m,
                Err(_) => continue,
            };

            let name = entry.file_name().to_string_lossy().to_string();
            let file_path = path.to_string_lossy().to_string();
            let is_dir = metadata.is_dir();
            let size = if is_dir { 0 } else { metadata.len() };
            let modified = metadata
                .modified()
                .map(system_time_to_string)
                .unwrap_or_default();

            files.push(FileInfo {
                name,
                path: file_path,
                is_dir,
                size,
                modified,
            });
        }
    }

    // Sort: directories first, then files, both alphabetically
    files.sort_by(|a, b| {
        match (a.is_dir, b.is_dir) {
            (true, false) => std::cmp::Ordering::Less,
            (false, true) => std::cmp::Ordering::Greater,
            _ => a.name.to_lowercase().cmp(&b.name.to_lowercase()),
        }
    });

    Ok(ApiResponse::ok(files))
}

/// Copy a file
#[tauri::command]
pub async fn copy_file(source: String, dest: String) -> Result<ApiResponse<()>, String> {
    match fs::copy(&source, &dest) {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(format!("复制文件失败: {}", e))),
    }
}

/// Check if path exists
#[tauri::command]
pub async fn path_exists(path: String) -> Result<ApiResponse<bool>, String> {
    Ok(ApiResponse::ok(PathBuf::from(&path).exists()))
}
